### PR TITLE
Search backend: mint SearchClient and use it for streaming API

### PR DESCRIPTION
**Pull request by camdencheek**
This creates a new `SearchClient` type and puts it to use in the streaming API. 

This PR is simultaneously a big deal because it completely cuts the dependency
between streaming search and `searchResolver`. The only dependencies on 
graphqlbackend for the streaming API are now `DecodedViewerFinalSettings`
and `LogSearchLatency`. I tried to pull `DecodedViewerFinalSettings` out
into a graphql-independent function, but it is very deeply ingrained and I'm not
up for that task right now.

## Test plan

Covered by integration tests with some manual testing to make me feel better.

### CODE REVIEW

File: internal/search/client/client.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	stream streaming.Sender,
+	inputs *run.SearchInputs,
+) (*search.Alert, error) {
+	jobArgs := &job.Args{
+		SearchInputs:        inputs,
+		Zoekt:               s.zoekt,
+		SearcherURLs:        s.searcherURLs,
+		OnSourcegraphDotCom: envvar.SourcegraphDotComMode(),
+	}
+	return execute.Execute(ctx, db, stream, jobArgs)
```
Comment (camdencheek) [inline internal/search/client/client.go:75]: I plan to inline both this and the call to `run.NewSearchInputs` since I don't really want those being called outside the context of `searchClient`. However, that can't happen until I modify `searchResolver` to use a `searchClient`, which I think makes sense to do in a separate PR.

File: internal/search/client/client.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+type SearchClient interface {
+	Plan(
+		ctx context.Context,
+		db database.DB,
+		version string,
+		patternType *string,
+		searchQuery string,
+		stream streaming.Sender,
+		settings *schema.Settings,
+	) (*run.SearchInputs, error)
```
Comment (camdencheek) [inline internal/search/client/client.go:29]: Note that this method signature is intentionally verbose. Some of these things won't be needed as we slim down what happens during the "plan" phase, at which point it will either be a more manageable set of args, or we'll decide to move the args into a dedicated input type. For now, this makes it easier to remove individual inputs.

File: cmd/frontend/graphqlbackend/search.go [hunk 3]
```diff
@@ -107,17 +107,17 @@ func (r *searchResolver) Inputs() run.SearchInputs {
 	return *r.SearchInputs
 }
 
-var mockDecodedViewerFinalSettings *schema.Settings
+var MockDecodedViewerFinalSettings *schema.Settings
 
-// decodedViewerFinalSettings returns the final (merged) settings for the viewer
-func decodedViewerFinalSettings(ctx context.Context, db database.DB) (_ *schema.Settings, err error) {
+// DecodedViewerFinalSettings returns the final (merged) settings for the viewer
+func DecodedViewerFinalSettings(ctx context.Context, db database.DB) (_ *schema.Settings, err error) {
```
Comment (camdencheek) [inline cmd/frontend/graphqlbackend/search.go:113]: Unfortunately, since we're no longer relying on resolvers, we now have to call this directly, which means exporting it. Not ideal, but it's a much narrower dependency than the whole `searchResolver` type.

File: internal/search/client/client.go (hunk 1 above)
Comment (camdencheek) [reply internal/search/client/client.go:75]: Also, this method has a slightly different signature than the `execute.Execute` signature because I want the zoekt and searcher clients to live statically on the search client, so we generate `jobArgs` on the fly here.

**Comment by camdencheek** [comment]
For some reason, integration tests are failing on CI, but seem to be working fine locally. And it's only streaming tests that are failing. It would make sense that this PR could break streaming things, but the fact that integration tests work locally is suspicious.

File: cmd/frontend/graphqlbackend/search_results_test.go [hunk 4]
```diff
@@ -104,8 +104,8 @@ func TestSearchResults(t *testing.T) {
 	t.Run("multiple terms regexp", func(t *testing.T) {
 		t.Skip("Skipping because it's currently failing locally")
 
-		mockDecodedViewerFinalSettings = &schema.Settings{}
-		defer func() { mockDecodedViewerFinalSettings = nil }()
+		MockDecodedViewerFinalSettings = &schema.Settings{}
+		defer func() { MockDecodedViewerFinalSettings = nil }()
```
Comment (rvantonder) [inline cmd/frontend/graphqlbackend/search_results_test.go:109]: oh look it's that test 😞 you'll have to rebase

**Review by rvantonder — APPROVED** [review]
ship-it
